### PR TITLE
Implement Dynamic Channel-Weighting MSE loss

### LINKED ISSUE #2485: Implement dynamically weighted MSE loss function
### Describe the task. It can be a feature, documentation, etc.

In Samudra 2 an improved MSE loss was introduced, using EMA to calculate channel weights dynamically during training. In their case it showed significant improvement in case of channels with low temporal variability. 

This loss function is easy to implement and worth having as an option 

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

https://arxiv.org/abs/2606.02610

### Area

- [ ] datasets, data readers, data preparation and transfer
- [x] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation
- [ ] performance

### PR DESCRIPTION
**Pull request by kacpnowak**
## Description

Implements the Dynamic Channel-Weighting Loss Function from the Samudra 2 paper to prevent variance collapse and imprinting artifacts in slow-evolving fields (like the deep ocean).

  Enable it inside  loss_fcts :
```yaml
      losses: {
        "physical": {
            type: LossPhysical,
            loss_fcts: {
                "mse": { },
                "dynamic_loss": { window: 100, L: 20.0 }
            },
        }
      }
```

## Issue Number

Closes https://github.com/ecmwf/WeatherGenerator/issues/2485

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [x] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

**Review by sophie-xhonneux — APPROVED** [review]
Looks good to me, shouldn't affect anybody if they are not using and backwards compatible as far as I can tell. @clessig your final call

**Review by clessig — COMMENTED** [review]
Thanks, interesting to see how this works for non-ocean but slowly evolving fields, e.g. soil moistures. 

The EMA functionality should be separated out into separate functions or a small class before merging.

**Review by clessig — CHANGES_REQUESTED** [review]

**Comment by kacpnowak** [comment]
@clessig thank your for the review. I've made the requested changes

**Comment by kacpnowak** [comment]
Before and after comparison is quite spectacular with 30 day forecast 
---
Before 

image

---

After
image

**Comment by clessig** [comment]
> ## Before and after comparison is quite spectacular with 30 day forecast
> Before
> 
> image
> After image

@kacpnowak : What does the grey area mean in the plot?

**Comment by kacpnowak** [comment]
Grey is ~7.5 degree Celsius. It's just so over smoothing that it predicts mean value almost everywhere

File: src/weathergen/train/loss_modules/loss_module_physical.py [hunk 1]
```diff
@@ -31,6 +31,55 @@ def get_num_samples(config) -> np.typing.NDArray:
...
+            weights_channels = weights_channels * weights_channels_static
+
+        return weights_channels
+
+    def update(self, stream_name: str, loss_lfct_chs: torch.Tensor):
+        if not self.enabled:
+            return
+
+        with torch.no_grad():
+            mse_per_chan = loss_lfct_chs.detach().clamp(min=1e-8)
```
Comment (clessig) [inline src/weathergen/train/loss_modules/loss_module_physical.py]: 1e-8 is order machine precision in fp32 (which is the highest precision we possibly have here). Shouldn't we clamp to a larger value?

File: src/weathergen/train/loss_modules/loss_module_physical.py [hunk 2]
```diff
@@ -57,6 +106,9 @@ def __init__(
         self.device = device
         self.name = "LossPhysical"
 
+        # Dynamic Loss state (extract it before parsing the actual loss functions)
+        self.dynamic_loss_cfg = loss_fcts.pop("dynamic_loss", None)
```
Comment (clessig) [inline src/weathergen/train/loss_modules/loss_module_physical.py]: Why do we need to pop here?

File: src/weathergen/train/loss_modules/loss_module_physical.py [hunk 3]
```diff
@@ -220,7 +278,18 @@ def compute_loss(self, preds: dict, targets: dict, metadata) -> LossValues:
 
             losses_all[stream_name] = defaultdict(dict)
 
-            stream_loss_weight, weights_channels = self._get_weights(stream_info)
+            stream_loss_weight, weights_channels_static = self._get_weights(stream_info)
+
+            if self.dynamic_loss_ema.enabled:
```
Comment (clessig) [inline src/weathergen/train/loss_modules/loss_module_physical.py]: Can we encapsulate this block also in `_get_weights()`

**Review by clessig — COMMENTED** [review]
Some minor comments.

File: src/weathergen/train/loss_modules/loss_module_physical.py (hunk 1 above)
Comment (kacpnowak) [reply src/weathergen/train/loss_modules/loss_module_physical.py]: I just set at lowest possible value with a plan to increase if loss gets unstable, but no such thing happened. Maybe it could be a configurable parameter?

File: src/weathergen/train/loss_modules/loss_module_physical.py (hunk 3 above)
Comment (kacpnowak) [reply src/weathergen/train/loss_modules/loss_module_physical.py]: Done

File: src/weathergen/train/loss_modules/loss_module_physical.py (hunk 2 above)
Comment (kacpnowak) [reply src/weathergen/train/loss_modules/loss_module_physical.py]: If was removed so it doesn't mess with other loss functions, but I changed it so it's `get` and it's checked later

**Review by clessig — APPROVED** [review]
Thanks.
